src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/Mock/StreamBuffer.cs

```diff
@@ -340,7 +340,10 @@ private void CancelWaiter()
 
             public void Reset()
             {
-                Debug.Assert(_hasWaiter == 0);
+                if (_hasWaiter != 0)
+                {
+                    throw new InvalidOperationException("Concurrent use is not supported");
```

Comment:
I'd suggest Debug.Fail'ing as well.  That'll fail fast, be less likely to be eaten, provide a better debugging experience (e.g. crash dump), etc.